### DISCUSSION TITLE
The redirects uses the default environment URL or need setting it?

### DISCUSSION THREAD

**Discussion by IgorDePaula** [nextauthjs/next-auth#2459] [Help]
### Question 💬

I push my app to aws amplify, I register, but on login, on got error the redirect action redirect me to localhost and not to provided url.

### How to reproduce ☕️

https://develop.d1cwc8lbezfn59.amplifyapp.com/   -> my app

my auth config
```
import NextAuth from 'next-auth'
import Providers from 'next-auth/providers'
import {Auth} from 'aws-amplify'

export default NextAuth({
    pages: {
        error: "/login", // Changing the error redirect page to our custom login page
    },
    callbacks: {
        async session(session, token) {
            return token
        },
    },
    providers: [
        Providers.Credentials({
            name: "Credentials",
            async authorize(credentials) {
                try {
                    const user = await Auth.signIn(credentials.email, credentials.password)
                    if (user) {
                        const data = {
                           name: user.attributes.name, email: user.attributes.email, image:'asdasdasd',id:user.attributes.sub,
                        }
                        return data
                    }
                } catch (e) {
                    const errorMessage = e.message
                    throw new Error(errorMessage + "&email=" + credentials.email)
                }
            },
        }),
    ],
})
```

Try you register and after do login.

(I'm using cognito, after register, the cognito will send a code for verification for your email, and you need send this code...after this you can login)

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
You probably forgot to set the `NEXTAUTH_URL` environment variable. It is required in production:

https://next-auth.js.org/configuration/options#nextauth_url

**Comment by IgorDePaula** [comment]
No, I put this variable
![Captura de Tela 2021-07-30 às 09 02 33](https://user-images.githubusercontent.com/2833993/127650262-03d2a888-d258-4938-a9b0-b3749552bacf.png)

**Comment by balazsorban44** [comment]
It looks like it's not being picked up for some reason. Try to log the value server-side. Also, unless you deployed the site on a subpath, the `NEXTAUTH_URL` should be the canonical URL (remove /api/path)

**Comment by IgorDePaula** [comment]
I tried alread, but no success.

**Comment by IgorDePaula** [comment]
The repository need upload any .env file?

**Comment by IgorDePaula** [comment]
I'm using next 11 and, the next auth config is error page is be login, but redirects to error page, why?

**Comment by balazsorban44** [comment]
Really hard to say. Do you have a reproduction?

This is certainly not a problem in `next-auth`, as far as I can tell.

.env is not necessary.

Read more about environment variables here https://nextjs.org/docs/basic-features/environment-variables

**Reply by IgorDePaula** [reply]
After exaustives tests, I noted that is not problem with next auth, sorry.

**Comment by recurser** [comment]
@balazsorban44 I’m having an identical issue after deploying to Amplify. The same app runs on Vercel and Heroku fine. The environment variable is set in Amplify, and the URL is canonical. Im trying to debug it at the moment, but was interested to find someone else here with the same issue.

**Reply by IgorDePaula** [reply]
try add  ```NEXTAUTH_URL``` on environment, on .env file. Amplify recognizes only this file, not from environment console.

**Reply by recurser** [reply]
@igordanchenko it looks like the Amplify environment variables only apply at **build time**, not at run time:

> Environment variables are key-value pairs that are available at build time ([source](https://docs.aws.amazon.com/amplify/latest/userguide/environment-variables.html))

I suspect we need to add the variables somewhere else (probably [in Lambda](https://docs.amplify.aws/guides/functions/configuring-lambda/q/platform/js#setting-an-environment-variable)?)

> Amplify recognizes only this file, not from environment console.

That file is available at runtime, which I suspect is why it works.

**Reply by IgorDePaula** [reply]
Who is @igordanchenko  ?

**Reply by dperrett** [reply]
@IgorDePaula sorry wrong name autompleted :)

**Reply by IgorDePaula** [reply]
ha ha ha, relax. 

```That file is available at runtime, which I suspect is why it works. ```

I tested, after write .env on repo, the redirect works fine.

**Comment by dperrett** [comment]
After a bit of digging I've got this working. The issue is that Amplify environment variables are only present at *build-time* within Amplify, and the `process.env.NEXTAUTH_URL` is required at *run-time* within Lambda.

Committing a `.env` file (as suggested by @IgorDePaula) is one solution - this works because the `.env` file ends up deployed to Lambda, where the variable is needed.

Another solution is to replace the `process.env.NEXTAUTH_URL` with the literal value during compilation within Amplify. A method for doing this is described [here](https://frontend-digest.com/environment-variables-in-next-js-9a272f0bf655). Basically, add an `env` section to `next.config.js` that will trigger replacement of the env-var with the literal value at build-time:

```
module.exports = {
  env: {
    NEXTAUTH_URL: process.env.NEXTAUTH_URL,
  },
  ...
};
```

@IgorDePaula's solution is probably simpler for non-sensitive vars such as `NEXTAUTH_URL`. If you need to pass sensitive vars such as API tokens the build-time replacement will probably be a better solution than committing the credentials to git.

**Reply by balazsorban44** [reply]
Nice! 👏
